### PR TITLE
New diag: unexpected '...'

### PR DESCRIPTION
**Pull request by isaacnonato**
#1023

### CODE REVIEW

**Comment by isaacnonato** [comment]
I have read and hereby agree to quick-lint-js' Contributor License Agreement (CLA-v1.md).

**Comment by isaacnonato** [comment]
recheck

**Review by strager — CHANGES_REQUESTED** [review]
Must fix: Please add test cases in test-parse-var.cpp and test-parse-expression.cpp.

In test-parse-var.cpp, test at least these cases:
```javascript
const [a, b, ...] = z;
const {a, b, ...} = z;
```

In test-parse-expression.cpp, test at least these cases:
```javascript
a = ...;
```

I think these two uses of `...` will need different diagnostic messages.

Let me know if you need help writing the tests.

File: src/quick-lint-js/fe/parse-expression.cpp [hunk 1]
```diff
@@ -114,11 +114,17 @@ void parser::visit_expression(expression* ast, parse_visitor_base& v,
     }
     break;
   }
+  case expression_kind::spread: {
+    // expression* child = this->parse_expression(v, precedence{.commas = false});
```
Comment (strager) [inline src/quick-lint-js/fe/parse-expression.cpp:118]: Nit: Delete unused code.
```suggestion

```

File: src/quick-lint-js/fe/parse-expression.cpp [hunk 2]
```diff
@@ -114,11 +114,17 @@ void parser::visit_expression(expression* ast, parse_visitor_base& v,
...
+    if (ast->child_0()->kind() == expression_kind::_missing) {
+      this->diag_reporter_->report(
+          diag_spread_must_precede_expression{this->peek().span()});
+    }   
+  }
```
Comment (strager) [inline src/quick-lint-js/fe/parse-expression.cpp:123]: Please do not implicitly fall through. Either write `[[fallthrough]];` or (better in this case) repeat the code:
```suggestion
    this->visit_expression(ast->child_0(), v, context);
  }
```

**Review by strager — APPROVED** [review]
Looks good to me. Let me know when I should merge your patch.

**Comment by isaacnonato** [comment]
Alright, I have made the suggested changes, I think it can be merged now.
Note: I used `[[fallthrough]];` because the other alternative failed some tests.

**Comment by strager** [comment]
Landed as commit ec5f3fca3412928b17e54de146872016c28f89aa. (I messed up ownership information for that commit; oops!)

**Comment by isaacnonato** [comment]
I think this also closes #932.

**Pull request closed without merging**